### ISSUE TITLE
Tracking Issue - Database table, column, index, and foreign keys are too long

### ISSUE THREAD

**Issue by derrickmehaffy** [strapi/strapi#13117]
### Describe the bug

This is a tracking issue for the various issues regarding lengths in the database being too long:

- `Error: ER_TOO_LONG_IDENT: Identifier name` - Foreign Keys: 
  - https://github.com/strapi/strapi/issues/12339
  - https://github.com/strapi/strapi/issues/13014
- `error: create index` - Index Names: 
  - https://github.com/strapi/strapi/issues/13031
- `Error: ER_TOO_LONG_IDENT: Identifier name` - Components 
  - https://github.com/strapi/strapi/issues/12101
  - https://github.com/strapi/strapi/issues/13116

We need to handle length constraints more gracefully across the entire database layer, maybe utilize short-codes like `comp`, ect.

### Steps to reproduce the behavior

See various linked issues

### Expected behavior

Each database should have properly length controlled tables, columns, indexes, and foreign keys

### Additional context

Additional duplicate issues will be linked in the comments below:

**Comment by derrickmehaffy** [comment]
https://github.com/strapi/strapi/issues/10636
https://github.com/strapi/strapi/issues/9199

https://forum.strapi.io/t/create-index-already-exists/16835/2

**Comment by terchris** [comment]
I see that the bug is classified as [severity: medium](https://github.com/strapi/strapi/labels/severity%3A%20medium) "If it breaks the basic use of the product but can be worked around"  But after this error strapi will not start and I cannot find a workaround to get it started again.

I guess I need to manually edit some files to get strapi started again. But what files and what to change ?

I'm upgrading from v 3.6.8 to 4.2.0 and I'm using the same names on fields and collections. (I tried to upgrade to v 4.3.0-beta1, but the same problem there. )

[2022-06-28 15:52:47.823] debug: :no_entry: Server wasn’t able to start properly.
[2022-06-28 15:52:47.826] error: create index “person_network_membership_roles_entity_newtork_membership_links_fk” on “person_network_membership_roles_entity_newtork_membership_links” (“person_network_membership_role_id”) - relation “person_network_membership_roles_entity_newtork_membership_links” already exists
error: create index “person_network_membership_roles_entity_newtork_membership_links_fk” on “person_network_membership_roles_entity_newtork_membership_links” (“person_network_membership_role_id”) - relation “person_network_membership_roles_entity_newtork_membership_links” already exists
at Parser.parseErrorMessage

**Comment by mckennapaul27** [comment]
Had the same issue.

Previously had collections called  'countryWalletVerticalLaw' and 'countryWalletGeneral' which made it fail.

Deleted the collections and renamed to 'CWVL' and 'CWG' and now it's working.

**Comment by Nikolaos-Kalfas** [comment]
> Deleted the collections and renamed to 'CWVL' and 'CWG' and now it's working.

Did you delete them from inside strapi or from the files? Because my strapi can't even start because of this.

**Comment by mckennapaul27** [comment]
My issue was it failed in production, it worked fine locally. Therefore I was able to delete the collections from within the strapi admin dashboard.

**Comment by PelirrojoAaron** [comment]
> I see that the bug is classified as [severity: medium](https://github.com/strapi/strapi/labels/severity%3A%20medium) "If it breaks the basic use of the product but can be worked around" But after this error strapi will not start and I cannot find a workaround to get it started again.
> 
> I guess I need to manually edit some files to get strapi started again. But what files and what to change ?
> 
> I'm upgrading from v 3.6.8 to 4.2.0 and I'm using the same names on fields and collections. (I tried to upgrade to v 4.3.0-beta1, but the same problem there. )
> 
> [2022-06-28 15:52:47.823] debug: ⛔ Server wasn’t able to start properly. [2022-06-28 15:52:47.826] error: create index “person_network_membership_roles_entity_newtork_membership_links_fk” on “person_network_membership_roles_entity_newtork_membership_links” (“person_network_membership_role_id”) - relation “person_network_membership_roles_entity_newtork_membership_links” already exists error: create index “person_network_membership_roles_entity_newtork_membership_links_fk” on “person_network_membership_roles_entity_newtork_membership_links” (“person_network_membership_role_id”) - relation “person_network_membership_roles_entity_newtork_membership_links” already exists at Parser.parseErrorMessage

For anyone else struggling with this, you need to delete the relational schema json that caused the CMS to crash, you will be able to find it in backend/src/components, that will prevent the schema from rebuilding the table on server start. At the moment the only way to get around this is to keep category and component names really short.

**Comment by GregorSondermeier** [comment]
Hello,

we're currently trying to migrate from v3 to v4. We do have some collectionNames/tables that are close to 64 chars in length. After using the [Codemods](https://github.com/strapi/codemods) to convert our models to the v4 format, when starting Strapi with an empty db, it fails because some table names, some column names or some foreign key identifiers are too long, because Strapi prepends a bunch of stuff (like `_links_inv_fk`) to the collectionNames.

For instance we get:

```
[2022-08-04 10:22:34.218] error: alter table `components_content_block_carousel_text_medias_content_links` add constraint `components_content_block_carousel_text_medias_content_links_inv_fk` foreign key (`content_block_id`) references `content_blocks` (`id`) on delete CASCADE - ER_TOO_LONG_IDENT: Identifier name 'components_content_block_carousel_text_medias_content_links_inv_fk' is too long
Error: ER_TOO_LONG_IDENT: Identifier name 'components_content_block_carousel_text_medias_content_links_inv_fk' is too long
    at Query.Sequence._packetToError (/home/gregor/repos/acker-platform/projects/bitbucket/ackerdemia/acker-cms-strapi-4/node_modules/mysql/lib/protocol/sequences/Sequence.js:47:14)
```

I understand that the 64 char limit is a restriction of SQL and not directly from Strapi. However it is still unfortunate.

We're now following the rocky path of first preparing our v3 database for the migration before executing the migration:
We shorten the collectionName in all our schemas (by coming up with a consistent naming pattern), rename the tables accordingly and hopefully come up with a script that finds and updates all references to the collectionName in the database.

**Comment by GregorSondermeier** [comment]
> and hopefully come up with a script that finds and updates all references to the collectionName in the database.

An awesome team member of mine did indeed come up with a script. We have now shortened all our v3 table names and can continue with the v4 migration.

**Comment by Krislunde** [comment]
@derrickmehaffy Any updates on this? I see the workaround is to create shorter collection names, but it still makes me anxious to use in production for clients.

**Comment by derrickmehaffy** [comment]
> @derrickmehaffy Any updates on this? I see the workaround is to create shorter collection names, but it still makes me anxious to use in production for clients.
> 
> Strapi will let you edit the displayname of components in the admin panel, but the `collectionName` field in component files stay the same, I assume in order to avoid data loss - leading to a lot of potentially error prone file editing/database work.

Currently no, I'm waiting on a response from our product and engineering teams. The workaround as you mentioned is to modify the collection name and use the manual migrations to rename the table during boot. You can see this WIP docs for the manual migrations: https://github.com/strapi/documentation/pull/889

**Comment by alphasldiallo** [comment]
> > Deleted the collections and renamed to 'CWVL' and 'CWG' and now it's working.
> 
> Did you delete them from inside strapi or from the files? Because my strapi can't even start because of this.

I ran into a similar issue and I wasn't able to access the Strapi admin dashboard. In my case, I had the following error: 
```
[2022-10-09 09:36:06.499] debug: ⛔️ Server wasn't able to start properly.
[2022-10-09 09:36:06.507] error: alter table `components_section_basic_section_with_buttons_components` add index `components_section_basic_section_with_buttons_component_type_index`(`component_type`) - ER_TOO_LONG_IDENT: Identifier name 'components_section_basic_section_with_buttons_component_type_index' is too long
Error: ER_TOO_LONG_IDENT: Identifier name 'components_section_basic_section_with_buttons_component_type_index' is too long
```
I fixed it by editing the file corresponding to the component causing the error: `basic-section-with-button`. The config file was stored under the path `src/components/section/basic-section-with-button.json`.
I changed the collection name and the display name inside the JSON file, saved the file and restarted the server. Now, it works like a charm.

**Comment by EarMaster** [comment]
I think it should be noted that the limits for SQLite are almost doubled. This means that - as many use SQLite during development - this first pops up when moving from development to staging or production. So I think the smallest limit should be enforced even if the current database allows longer identifiers.

**Comment by derrickmehaffy** [comment]
> I think it should be noted that the limits for SQLite are almost doubled. This means that - as many use SQLite during development - this first pops up when moving from development to staging or production. So I think the smallest limit should be enforced even if the current database allows longer identifiers.

Indeed, this is something that we will have to wait until v5 to change as it would be considered breaking (enforcing) but we do need to accurately determine limits including any automated prefix/suffix/etc and then enforce that in the CTB.

We can't stop users trying to exceed that limit manually through the schema files but I know our DX squad (cc @marcoautiero ) does intend to have a period here soon where we work on the CTB and make some sane improvements.

**Comment by EarMaster** [comment]
I disagree. This behaviour currently breaks the whole setup on deployment and requires fundamental code changes late in the development process. There should *at least* be warnings that output in development mode that db identifiers may be to long for certain database types.

**Comment by EarMaster** [comment]
You didn't say you plan to warn users. Good to hear that you agree.

I also encourage you to note these limits in the documentation as it will be breaking bug for some users during deployment.

**Comment by paquettea** [comment]
A workaround I've seen in drupal, is that it creates an hash of the resource name when it's too long for the database. In the end it could generate something like components_fd361487_index from components_section_basic_section_with_buttons_component_type_index where fd361487 is a Adler32 hash of the stripped part (section_basic_section_with_buttons_component_type)

It is less convenient when looking at the database, but it's better than having the solution breaking down on something that is already abstracted for the dev.

I've encountered the same issue moving from strapi 3.6 with mongoose to strapi 4 with mysql.

**Comment by antokhio** [comment]
Hi just came across this:
```py
error: create index "components_notifications_collaboration_sents_collaboration_links_fk" on "components_notifications_collaboration_sents_collaboration_links" ("collaboration_sent_id") - relation "components_notifications_collaboration_sents_collaboration_link" already exists
error: create index "components_notifications_collaboration_sents_collaboration_links_fk" on "components_notifications_collaboration_sents_collaboration_links" ("collaboration_sent_id") - relation "components_notifications_collaboration_sents_collaboration_link" already exists```

Had to manually rename:
```json
{
   "collectionName": "components_notifications_collaboration_accepteds",
 }
 ```
 to
 ```json
{
   "collectionName": "components_notifications_collab_accepteds",
 }
 ```
 
 this breaks server without way to restart unless you mod/remove component by hand
- Postgres:14
- Strapi 4.5.0

**Comment by derrickmehaffy** [comment]
Prod/Eng team, raising this to high as it impacts some EE customers as well.

**Comment by GregorSondermeier** [comment]
Soo.. When upgrading from v4.4.7 to 4.5.6 in order to address the security issue this popped up *again*:
![tmp_61b03c2a-e804-4dcd-a54a-ae9956736d34~2](https://user-images.githubusercontent.com/11959747/213470268-4e24d776-4943-475e-b70b-bdbd83848730.png)
I know, the names "entities_collection" and "wiki_articles_collection" are rather stupid, but they are given in the project I have inherited.

Now we have to change the names and migrate the content before we can upgrade to the latest Strapi version. Again.

**Comment by awan1** [comment]
Do folks have a clear idea of how long `{componentGroup}_{componentName}` can be? I think this depends on the longest suffix, which looks like `links_order_fk`, and `component_{group}_{componentName}_links_order_fk` must be at most 64char, which means that {group}_{componentName} must be at most 64-25=39char?

**Comment by GregorSondermeier** [comment]
> This limits `{group}_{content_name_plural}` to 32char.

Uff that's not very much. As soon as you group components within a folder, you'll run intro trouble sooner or later.

**Comment by derrickmehaffy** [comment]
> > This limits `{group}_{content_name_plural}` to 32char.
> 
> Uff that's not very much. As soon as you group components within a folder, you'll run intro trouble sooner or later.

We are going to need to define shorter prefixes and what not  🤔

**Comment by gwynhokkers** [comment]
Just encountered this exact issue today so still seems to be an issue.

- Strapi v4.6.1
- mysql

**Comment by ppemond-bkom** [comment]
We just encountered the issue as well... Would be nice to have a clear documentation on what we should do in this case. Thank you

**Comment by christianheine** [comment]
Same here. We tried to upgrade from v3 to v4 and now essentially got stuck at the database migration.

Manually renaming long components and collections seems to allow the initial database creation (e.g., renaming components from "components_primitives_bundling_options" to "components_primitives_bo"), but then we can't use the data migration scripts (unless we figure out a way to migrate the v3 database). 

And on top of the migration issue it gets harder to add new components with descriptive names. Because when creating new components inside a given category the internal naming cannot be controlled (at least when using the UI). For example, when creating a new component called "test-component" inside "primitives" it is automatically created as "components_primitives_test_components".

If the internal names are not exposed to the user, maybe some sort of hashing (or uids) would be safer? Or at least have a way to control how internal names are set.

Though for the short term even a guide on how to prepare the v3 database (and schema) for a migration would be really highly appreciated.

**Comment by TheBit** [comment]
@derrickmehaffy 
> We are going to need to define shorter prefixes and what not 🤔

Here is what our typical code-review process looks like when reviewing changes in Strapi:
![alt](https://i.gyazo.com/ba38f3ee485ab8060926d5514f90271d.png)

We just created and follow next convention (get rid from folder name and shorten `components` to `cmp`):
```
components_{comp-folder}_{comp-name} => cmp_{comp-name}
```

**Comment by derrickmehaffy** [comment]
We are planning on looking into this in v5. For now we cannot change the table name structure in v4 as it would be considered a massive breaking change.

Things like shortening the linking terms and hashing are not off the table 😉 but we can't do it now.

**Comment by derrickmehaffy** [comment]
Raising sev from high to critical as this is impacting self-hosted EE customers and cloud users as per our EE sev matrix internally.

![image](https://user-images.githubusercontent.com/8593673/235977854-33feefa3-092b-4283-bc49-6acc3accb8c3.png)

**Comment by kepek** [comment]
Any ETA?

**Comment by derrickmehaffy** [comment]
> > I think it should be noted that the limits for SQLite are almost doubled. This means that - as many use SQLite during development - this first pops up when moving from development to staging or production. So I think the smallest limit should be enforced even if the current database allows longer identifiers.
> 
> Indeed, this is something that we will have to wait until v5 to change as it would be considered breaking (enforcing) but we do need to accurately determine limits including any automated prefix/suffix/etc and then enforce that in the CTB.
> 
> We can't stop users trying to exceed that limit manually through the schema files but I know our DX squad (cc @marcoautiero ) does intend to have a period here soon where we work on the CTB and make some sane improvements.

@kepek see my earlier response, it won't be possible to fix in v4 as it would be a breaking change and is something we plan to include in v5 but that will not have a stable release in 2023.

**Comment by GregorSondermeier** [comment]
If you're stuck migrating within Strapi v4 (say v4.4 to v4.5) because of this issue, you'd need to adjust your content type schemas. But by default, you will experience a data loss when Strapi starts next time, because it syncs the tables with your schemas and deletes everything that is not defined by your schemas.

To circumvent this data loss, you could create a database migration script which syncs the tables with your schemas. This feature is documented here:
https://docs.strapi.io/dev-docs/database-migrations

Strapi makes sure that the database migration script is executed _once between_ starting Strapi and syncing the tables with the content type schemas. This is done by storing the executed migrations in the __strapi_migrations_ table.

![image](https://github.com/strapi/strapi/assets/11959747/976cd1db-3d36-4e94-82e9-9a6db1b38d45)

**Comment by BarrySaikSoundry** [comment]
I'm running into this as well on v4. It's really limiting and easy to fall into, since it's very tempting to create nice folder names to organize your content but you quickly use up all the available characters out of the 64 characters Postgres will allow for the table name.  I have a component folder for general components and I get this: components_components_headline_with_subheads

**Comment by gorlitzer** [comment]
Hello there. Any news on this? Running `strapi 4.12.4` still having issue

**Comment by derrickmehaffy** [comment]
Again see: https://github.com/strapi/strapi/issues/13117#issuecomment-1550383637

Locking this, we can't fix it in v4, it will be addressed in v5

**Comment by innerdvations** [comment]
@derrickmehaffy This is solved in the v5 betas, and unsolvable without major breaking changes in v4. Should we finally close this issue?

**Comment by derrickmehaffy** [comment]
> @derrickmehaffy This is solved in the v5 betas, and unsolvable without major breaking changes in v4. Should we finally close this issue?

Indeed I think it's probably about time.

Marking as resolved (finally) in Strapi 5
